### ISSUE TITLE
Check for mis-ordered usage of errors (effectively ignored errors while technically being used out of order)

### ISSUE THREAD

**Issue by curquhart** [kisielk/errcheck#260]
I'm not 100% sure if this case would fall under the errcheck linter (since err IS being used, just not where one would expect it to be used) but i wanted to check if it would be considered an issue (if so and if you're into it i'll look into writing a fix)

fixing this would for sure be a breaking change (producing false positives) since while in my example, it will clearly panic, it's obviously much less cut and dry in (massive amounts of) real world code, so doing so might require a flag?

<details>
<summary>Reproducible example which passes errcheck</summary>

```
package main

import "fmt"

func foo() (*struct{ Foo string }, error) {
	return nil, fmt.Errorf("foo")
}

func main() {
	var err error
	println(err)
	res, err := foo()

	if res.Foo == "" { // this will ofc panic
		println("foo")
	}
}
```
</details>

this specific example does get caught by the ineffassign linter if we're running golangci-lint, but i believe it should be caught by errcheck. if we add usage of `err` after the res.Foo check, it passes both linters (but I would also expect this case to fail errcheck as well)

<details>
<summary>Reproducible example which passes both errcheck and ineffassign (for good measure)</summary>

```
package main

import "fmt"

func foo() (*struct{ Foo string }, error) {
	return nil, fmt.Errorf("foo")
}

func main() {
	res, err := foo()

	if res.Foo == "" { // this will ofc panic
		println("foo")
	}
	println(err)
}
```
</details>

thanks!

**Comment by dtcaciuc** [comment]
@kisielk and/or @echlebek might correct me, but I would say right now errcheck scope of responsibility is basically limited to checking whether the author recognizes that function/method returns an error by explicitly assigning it to something. 

For everything else, we would need to introduce a much more complex analysis. Like for instance, in your example the "// this will ofc panic" is only ofc if errcheck somehow understands that foo() returns a nil which requires analyzing the called function. It's also possible that a partial meaningful value is returned even though an error is returned alongside it and the developer may choose to do something with it and handle error later (such as log its contents). 

Considering that other tools already address some of this, I would be reluctant to try and introduce that into errcheck.

**Comment by curquhart** [comment]
that's fair! i figured it may or may not make sense here.

thanks 😊

**Comment by curquhart** [comment]
i didn't, unfortunately

**Comment by ccoVeille** [comment]
cc @alingse who loves challenges

**Comment by echlebek** [comment]
@curquhart @danielhochman this use case is solved by staticcheck (https://staticcheck.dev/). We consider staticcheck to be superior to errcheck, which is only maintained for legacy purposes now.
